### ISSUE TITLE
Enabling LTO produces broken extension modules

### ISSUE THREAD

**Issue by ra-kete** [Nuitka/Nuitka#930]
* Nuitka version, full Python version and Platform (Windows, OSX, Linux ...)
```
$ python -m nuitka --version
0.6.10.5
Python: 3.6.9 (default, Oct  8 2020, 12:12:24)
Executable: /tmp/nuitka-bug/.venv/bin/python
OS: Linux
Arch: x86_64
```

* How did you install Nuitka and Python (pip, anaconda, deb, rpm, from source,
  what is a virtualenv ...), this is very important usually.

Nuitka was installed with `pip install nuitka`.

* If possible please supply a [Short, Self Contained, Correct, Example](http://sscce.org/)
  that demonstrates the issue i.e a small piece of code which reproduces
  the issue and can be run with out any other (or as few as possible)
  external dependencies.

This can be reproduced by creating a single empty Python file, building it into an extension module with LTO enabled and trying to import the extension module:
```
$ touch test.py

$ python -m nuitka --module test.py --output-dir=build --lto
Nuitka:INFO: Starting Python compilation.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating C source code for backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C level backend compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: gcc (gcc).
Nuitka-Scons:WARNING: You are not using ccache.
Nuitka:INFO: Successfully created 'build/test.cpython-36m-x86_64-linux-gnu.so'.

$ cd build && python -c "import test"
Missing decoding for 0

Aborted
```

Importing the extension module produces an unhelpful error message and aborts the interpreter. Expected behavior would be no output in this case.

If the extension module is compiled without the `--lto` flag, it can be imported without issues.

Also note that I seem to be able to compile other (larger) projects with LTO just fine. It looks like the issue only shows for some projects, but I couldn't find any reason for this.

* If this is a regression (used to work in an earlier version of Nuitka),
  please note what you know about that.

This issue did not exist in Nuitka versions prior to 0.6.9. As far as I can tell, 0.6.8.4 was the last version that made the above minimal example not crash.

**Comment by ra-kete** [comment]
This might be related to #868 as we see the same error message there. Though Conda is not involved in this case.

**Comment by kayhayen** [comment]
That's a big oops for LTO on Linux right there.

I don't think it's related to the mentioned issue, just the same symptom. The constants blob is being added through an object file, with `ld` directly, and it seems to treat my blob unfairly. I reproduced this on my Debian just now the same. Binaries and non-lto work both well as you report.

The constants blob is not new by itself, but the new code introduced in 0.6.9 might be more sensitive to noticing corruptions. I will have to read up on this. Forcing to use C code for the constants blob data makes it work. Obviously going via the C compiler for data is not good for (ccache) caching, nor any good for compile times, but probably also not too bad.

**Comment by kayhayen** [comment]
Apply this mini-diff as a workaround:

```
-    and getLinkerArch(target_arch=target_arch, mingw_mode=mingw_mode) is not None
+    and getLinkerArch(target_arch=target_arch, mingw_mode=mingw_mode) is not None and not lto_mode
```

**Comment by kayhayen** [comment]
I also tried clang, but it behaves the same, although I think it's using a different linker.

**Comment by kayhayen** [comment]
```
// We need to make sure this is local to the package, or else it will
// be taken from any external linkage.
#if defined(_NUITKA_MODULE) && !defined(_NUITKA_CONSTANTS_FROM_RESOURCE) && !defined(__cplusplus)
const unsigned char constant_bin[0];
#endif
```

There is also this, which I am myself confused about now. I believe the symbol ought to not be exported, so this wouldn't have a point, removing that also seems to make it work, so the question becomes why it got added. :)

**Comment by kayhayen** [comment]
So, the commit that introduced this first was done for Fedora 32 which apparently found something wrong with it, and that is a workaround that can't quite last. I am assuming the LTO then has a duplicate symbol, where this one wins, and it should not.

**Comment by kayhayen** [comment]
Yes, I think I am seeing that `constant_bin` becomes a symbol to be exported, which can't be right, if that declaration is removed.

I guess, I need to convince the linker to include an object, but without exporting it there.

**Comment by kayhayen** [comment]
I found this, it seems to do everything but MSVC, which is fine, we can use the resource mode there too:

https://github.com/graphitemaster/incbin

It seems this way, the inline assembly is used and a binary blob include can be done there. I will need to convince myself if the symbol is exported, but I bet, that can be addressed.

**Comment by kayhayen** [comment]
This is fantastic, it seems to be already working, and it's only less code, esp. it doesn't need to know linker arch, and should be equally fast. It ought to work on non-Linux as well, probably allowing me to remove the source code mode for constants.

**Comment by kayhayen** [comment]
And on the upside, the created binary is smaller, the symbol becomes global, but this might be something patchable. I am going to submit this to my private and public CI, lets see what comes up. Fedora 33 didn't show the problem of duplicate symbols, but the OBS build with Fedora 32 might, but I doubt it can happen this way.

**Comment by kayhayen** [comment]
The Windows gcc (MinGW64) works with it nicely, clangcl.exe does not, seemingly too MSVC compatible there, I didn't try the normal clang.exe yet on Windows yet, but I doubt it can be an issue. I noticed though, that for some reason on Windows the resource mode binary is smaller, which is strange, so I am not using it there yet.

**Comment by kayhayen** [comment]
In the CI, on RHEL 6 and older, the constants blob wasn't working anymore for some test cases, which is strange. I have therefore made a gcc version check for `>= 7` to enable it now, but I am also adding a verfication at run time, taking a CRC32 of the constants blob, to see if it has been tampered with. That should catch #868 and alikes sooner, as well as this one. 

Not sure if what I have seen there is a incbin bug, it definitely is one if the constants binary mismatches for some cases, but for now, I am aiming at getting it to work for most cases, continuing to use the old code with old gcc, but then it would surely be nice to get to use incbin with everything.

**Comment by kayhayen** [comment]
I had to update it again, since incbin is leading to exported symbols, for 2 extension modules, the second one would load the wrong one. Got around that by adding a local feature for incbin, now it's working.

Also where it wasn't working, it seems that an existing bug was caught and that got fixed too. For case collisions of module names Nuitka was using seemingly the wrong constants blob on case sensitive file systems. Not sure really how this worked befoe, or why it wasn't occuring before, but it had nothing to do with incbin at all.

So incbin is now used for everything but Windows, where resources are still used.

There is an impact of the C code that does the incbin not being cacheable and a need to pass the valiue via define. I guess, the code to get the data included ought to be generated instead, and be a separate module, then this won't be an issue and there is no need to pass anything.

**Comment by kayhayen** [comment]
I changed it again, generating a small C file that will be uncachable. Newer ccache detects that, but older on e.g. RHEL 6 does not, so this is actively disabled for just that one file, which is a good idea anyway, but it cost me really a lot of time to debug why it was using corrupt constant blobs, that had good CRC32 checks.

Anyway, the improvement is real and will be in the next release that is bound to happen any minute now, running final pre-release tests to see what has issues before making actual release.

**Comment by kayhayen** [comment]
Released as part of 0.6.11 just now.
